### PR TITLE
Improve chat aa1 to master

### LINKED ISSUE #56: Improve group chat functionality (adding people)
**Is your feature request related to a problem? Please describe.**
Currently it is not possible to add people to a group chat.

**Describe the solution you'd like**
It should be possible to add people to a group chat.

### LINKED ISSUE #65: Focus on chat when switching
**Describe the feature or change you'd like**
When switching a chat the text box should automatically focus. This could be called in componentDidMount React Lifecycle method.

### PR DESCRIPTION
**Pull request by Theralun**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently message field is not selected by default. #65 
- Currently there is no way of adding people to group chats. #56
- Currently you cannot leave group chats.
- Currently you cannot name a group chat any name with "and".
- Currently there is no way to start a chat from the profile page.
- Currently you cannot switch chats after viewing one via redirect.
- Currently the chat name is not displayed on top of the chat.
- Currently unread chats are not shown outside of the chat page.

**What is the new behavior (if this is a feature change)?**
- The message field is now selected when loading a chat. Closes #65 
- Now you can add users to group chats. Closes #56
- There is now a function to leave group chats.
- Now names with "and" in them are permitted.
- There is now a button to start chats when clicking on a username in the profilepage (Chats and Posts)
- Changing of chats after clicking a notification has been fixed.
- The chat name is now displayed on top of the chat component.
- A badge for unread chats has been added to the navigation.

**Other information**:
- Minor version update to `v3.3.0`.

### CODE REVIEW

File: src/components/molecules/ReceivedUser/index.jsx [hunk 1]
```diff
@@ -29,6 +35,24 @@ import "./receiveduser.scss";
 
 //#region > Components
 class ReceivedUser extends React.Component {
+  startChat = async () => {
+    console.log(this.props.profile, this.props.receivedUser.uid);
```
Comment (Aichnerc) [inline src/components/molecules/ReceivedUser/index.jsx:39]: Yeet this crap out of here pls.

File: src/components/organisms/Chat/index.jsx [hunk 2]
```diff
@@ -141,11 +174,250 @@ class Chat extends React.Component {
...
+    const { selectedUsers } = this.state;
+
+    const selectedUserIDs = selectedUsers.map((u) => {
+      return u.id;
+    });
+
+    this.props.joinChat(selectedUserIDs, chatDetails.id, chatDetails.users);
+
+    let msgString = "";
+    console.log(selectedUsers);
```
Comment (Aichnerc) [inline src/components/organisms/Chat/index.jsx:204]: Bruh. Yeet this blind, deaf, comatose, lobotomy patient out of here.

File: src/components/molecules/ReceivedUser/index.jsx (hunk 1 above)
Comment (Theralun) [inline src/components/molecules/ReceivedUser/index.jsx:39]: Yeet dis

File: src/components/organisms/Chat/index.jsx (hunk 2 above)
Comment (Theralun) [inline src/components/organisms/Chat/index.jsx:204]: Yeet dis

File: src/components/pages/ChatPage/index.jsx [hunk 3]
```diff
@@ -316,7 +329,10 @@ class ChatPage extends React.Component {
     // Check users are present
     if (name && users.length > 1) {
       // Check if the group name has more than 2 characters
-      if (name.length > 2) {
+      if (
+        name.length > 2 &&
+        !name.split(" ").includes(process.env.REACT_APP_ACTION_CHAT_BINDER)
```
Comment (Theralun) [inline src/components/pages/ChatPage/index.jsx:334]: hälp

```suggestion
        if( item.chat.name.split( process.env.REACT_APP_ACTION_CHAT_BINDER ).length === 2 )
```

File: src/store/actions/chatActions.js [hunk 4]
```diff
@@ -1,13 +1,28 @@
...
+      .then((result) => {
+        return { id: result.id, data: result.data() };
+      });
 
+    users.forEach((u) => {
+      if (!currentData.data.users.includes(u)) {
+        usersToAdd = [...usersToAdd, u];
+      }
+    });
     // Add user to the chat users
```
Comment (Theralun) [inline src/store/actions/chatActions.js:21]: New line before

File: src/store/actions/chatActions.js [hunk 5]
```diff
@@ -17,6 +32,46 @@ export const joinChat = (uid, chid, curUsers) => {
...
+
+    // Get chat from collection
+    const currentData = await firestore
+      .collection("chats")
+      .doc(chatDetails.id)
+      .get()
+      .then((result) => {
+        return { id: result.id, data: result.data() };
+      });
+    console.log(currentData);
```
Comment (Theralun) [inline src/store/actions/chatActions.js:47]: yeet

File: src/store/actions/chatActions.js [hunk 6]
```diff
@@ -17,6 +32,46 @@ export const joinChat = (uid, chid, curUsers) => {
...
+
+    let newUsers = currentData.data.users.filter((u) => u !== uid);
+    console.log(newUsers);
```
Comment (Theralun) [inline src/store/actions/chatActions.js:50]: yeet

File: src/components/pages/ChatPage/index.jsx (hunk 3 above)
Comment (Theralun) [reply src/components/pages/ChatPage/index.jsx:334]: Not needed anymore

**Pull request merged**